lib/src/config_document_parser.cc

```diff
@@ -0,0 +1,632 @@
+        if (t->get_token_type() == token_type::UNQUOTED_TEXT) {
+            string kind_text = t->token_text();
+
+            config_include_kind kind;
+            if (kind_text == "url(") {
+                kind = config_include_kind::URL;
+            } else if (kind_text == "file(") {
+                kind = config_include_kind::FILE;
+            } else if (kind_text == "classpath(") {
+                kind = config_include_kind::CLASSPATH;
+            } else {
+                throw parse_exception("expecting include parameter to be quoted filename, file(), classpath(), or "
+                                              "url(). No spaces are allowed before the open paren. Not expecting: " +
+                                                t->to_string());
+            }
+
+            children.push_back(make_shared<config_node_single_token>(t));
+
+            // skip space inside parens
+            t = next_token_collecting_whitespace(children);
+
+            // quoted string
+            if (!tokens::is_value_with_type(t, config_value_type::STRING)) {
+                throw parse_exception("expecting a quoted string inside file(), classpath(), or url(), rather than" +
+                t->to_string());
+            }
+            children.push_back(make_shared<config_node_simple_value>(t));
+
+            // skip space inside parens
+            t = next_token_collecting_whitespace(children);
+
+            if (t->token_text() != ")") {
+                throw parse_exception("expecting a close parentheses ')' here, not: " + t->to_string());
+            }
+            return make_shared<config_node_include>(children, kind);
+        } else if (tokens::is_value_with_type(t, config_value_type::STRING)) {
+            children.push_back(make_shared<config_node_simple_value>(t));
+            return make_shared<config_node_include>(children, config_include_kind::HEURISTIC);
+        } else {
+            throw parse_exception("include keyword is not followed by a quoted string, but by: " + t->to_string());
+        }
+    }
+
+    shared_ptr<config_node_complex_value> parse_context::parse_object(bool had_open_curly) {
+        // invoked just after the OPEN_CURLY (or START, if !hadOpenCurly)
+        bool after_comma = false;
+        path* last_path = nullptr;
+        bool last_inside_equals = false;
+        shared_node_list object_nodes;
+        shared_node_list key_value_nodes;
+        unordered_map<string, bool> keys;
+
+        if (had_open_curly) {
+            object_nodes.push_back(make_shared<config_node_single_token>(tokens::open_curly_token()));
+        }
+
+        while (true) {
+            shared_token t = next_token_collecting_whitespace(object_nodes);
+            if (t->get_token_type() == token_type::CLOSE_CURLY) {
+                if (_flavor == config_syntax::JSON && after_comma) {
+                    throw parse_exception(add_quote_suggestion(t->to_string(),
```

Comment:
This signature for add_quote_suggestion happens often enough it'd be nice to still have the shorter call signature.